OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,346 @@
+			connectAutomaticallyCheckBox.OnClick = () => Game.Settings.Irc.ConnectAutomatically ^= true;
+
+			var connectButton = connectBG.Get<ButtonWidget>("CONNECT_BUTTON");
+			connectButton.OnClick = () =>
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+
+				Game.Settings.Irc.ConnectAutomatically = connectAutomaticallyCheckBox.IsChecked();
+				Game.Settings.Save();
+				Connect();
+			};
+		}
+
+		void Connect()
+		{
+			client = new IrcClient();
+			client.Encoding = System.Text.Encoding.UTF8;
+			client.SendDelay = 100;
+			client.ActiveChannelSyncing = true;
+
+			client.OnConnected += OnConnected;
+			client.OnError += OnError;
+			client.OnRawMessage += OnRawMessage;
+			client.OnJoin += OnJoin;
+			client.OnChannelActiveSynced += OnChannelActiveSynced;
+			client.OnNickChange += OnNickChange;
+			client.OnPong += OnPong;
+
+			try
+			{
+				AddChatLine("Connecting to {0}...".F(Game.Settings.Irc.Hostname));
+				client.Connect(Game.Settings.Irc.Hostname, Game.Settings.Irc.Port);
+			}
+			catch (Exception e)
+			{
+				AddChatLine("Connection error: {0}".F(e.Message));
+				Log.Write("irc", e.ToString());
+			}
+
+			listenThread = new Thread(new ThreadStart(Listen));
+			listenThread.Name = "IrcListenThread";
+			listenThread.Start();
+		}
+
+		void OnPong(object sender, PongEventArgs e)
+		{
+			if (pingSent)
+			{
+				AddChatLine("PONG recieved after {0} ms.".F(e.Lag.Milliseconds));
+				pingSent = false;
+			}
+			else
+				Log.Write("irc", "PONG sent after {0} ms.".F(e.Lag.Milliseconds));
+		}
+
+		public void Listen()
+		{
+			client.Listen();
+		}
+
+		Widget MakeLabelWidget(LabelWidget template, string item)
+		{
+			var widget = (LabelWidget)template.Clone();
+			var font = Game.Renderer.Fonts[widget.Font];
+			item = WidgetUtils.WrapText(item, widget.Bounds.Width, font);
+			widget.Bounds.Height = font.Measure(item).Y;
+			widget.GetText = () => item;
+			return widget;
+		}
+
+		void AddChatLine(string text)
+		{
+			Log.Write("irc", text.Replace("{", string.Empty).Replace("}", string.Empty));
+			Game.RunAfterTick(() =>
+			{
+				var scrolledToBottom = historyPanel.ScrolledToBottom;
+
+				var newChild = MakeLabelWidget(historyTemplate, text);
+				historyPanel.AddChild(newChild);
+
+				if (scrolledToBottom)
+					historyPanel.ScrollToBottom(smooth: true);
+			});
+		}
+
+		bool EnterPressed()
+		{
+			if (!inputBox.Text.Any())
+				return true;
+
+			var text = inputBox.Text;
+			inputBox.Text = "";
+
+			if (text.StartsWith("/nick "))
+			{
+				var nick = text.Replace("/nick ", string.Empty);
+				if (Rfc2812.IsValidNickname(nick))
+					client.RfcNick(nick);
+				else
+					AddChatLine("Invalid nickname.");
+			}
+			else if (text.StartsWith("/ping "))
+			{
+				client.RfcPing(client.GetIrcUser(text.Replace("/ping ", string.Empty)).Host);
+				pingSent = true;
+			}
+			else if (text.StartsWith("/"))
+				AddChatLine("Unknown command");
```

Comment:
@Unit158: Players shouldn't even *know* that it is using IRC behind the scenes.  There's no reason for them to try to use IRC commands, so no need to warn them.